### ISSUE TITLE
Support status for vite-wasm-plugin and testing env

### ISSUE THREAD

**Issue by Jerboas86** [Menci/vite-plugin-wasm#24]
I was trying to uderstand where vite-wasm-plugin is supported.  
The testing environment is here: [https://github.com/Jerboas86/vpw-test-env](https://github.com/Jerboas86/vpw-test-env)

I tried different frameworks:

- svelte (CSR)
- sveltekit (SSR)

In different scopes:

- Window scope
- Worker scope
- Worklet scope (audioworklet)
- Nodejs scope (vitest)

And with different web browers:

- Chrome 108
- Firefox 108
- Gnome web (aka Epiphany) 43

### Chrome

|         | Svelte | Sveltekit |
| ------- | ------ | --------- |
| Window  | ✅     | ❗        |
| Worker  | ✅     | ❗        |
| Worklet | ❗     | ❗        |
| Vitest  | ❗     | ❗        |

### Firefox

|         | Svelte | Sveltekit |
| ------- | ------ | --------- |
| Window  | ✅     | ❗        |
| Worker  | ❗     | ❗        |
| Worklet | ❗     | ❗        |
| Vitest  | ❗     | ❗        |

### Gnome web

|         | Svelte | Sveltekit |
| ------- | ------ | --------- |
| Window  | ✅     | ❗        |
| Worker  | ✅     | ❗        |
| Worklet | ❗     | ❗        |
| Vitest  | ❗     | ❗        |

opened issues:
- [ Firefox support for web workers ](https://github.com/Menci/vite-plugin-wasm/issues/18)
- [SSR support] (https://github.com/Menci/vite-plugin-wasm/issues/4)

and opened PR:
- [vitest support](https://github.com/Menci/vite-plugin-wasm/pull/16)

I would suggest to:
- update the readme to reflect current support
- and open issue for audioworklet support

The testing can also be extended if it is valuable, let me known.

**Comment by Menci** [comment]
<img width="1329" alt="image" src="https://user-images.githubusercontent.com/11341955/216823675-60457962-f274-49f1-9f7f-689d89f2d7be.png">

Can't start your project

**Comment by Menci** [comment]
I managed to build your project and added Vitest and SSR support. Please test~

**Comment by Menci** [comment]
A question: what's the difference between Worker and Worklet. What extra steps will I need to make it work in Worklet?

**Comment by Jerboas86** [comment]
With vite-plugin-wasm: v3.2.1

## Results

### Vitest

|        | Svelte | Sveltekit |
| ------ | ------ | --------- |
| Vitest | ✅     | ✅        |

### Chrome

|         | Svelte | Sveltekit |
| ------- | ------ | --------- |
| Window  | ✅     | ✅        |
| Worker  | ✅     | ✅        |
| Worklet | ❗     | ❗        |

### Firefox

|         | Svelte | Sveltekit |
| ------- | ------ | --------- |
| Window  | ✅     | ✅        |
| Worker  | ❗     | ❗        |
| Worklet | ❗     | ❗        |

### Gnome web

|         | Svelte | Sveltekit |
| ------- | ------ | --------- |
| Window  | ✅     | ✅        |
| Worker  | ✅     | ✅        |
| Worklet | ❗     | ❗        |

**Comment by Jerboas86** [comment]
> A question: what's the difference between Worker and Worklet. What extra steps will I need to make it work in Worklet?

AFAIK, In AudioWorkletScope:
- `URL` is not available.
- `fetch` is not available.
- `import` is not available (yet)  in Firefox's worklet [(see)](https://bugzilla.mozilla.org/show_bug.cgi?id=1572644)
  
Other apis are not available [(see)](https://github.com/WebAudio/web-audio-api/issues/2499)

I dont known if this is also true for others worklet: PainWorklet, LayoutWorklet, and AnimationWorklet.

**Comment by Menci** [comment]
Does it mean I could only base64 encode the WASM file to load it in Worklet context? Are there any other APIs used by wasm-bindgen that is missing in Worklet context?

**Comment by Jerboas86** [comment]
> Does it mean I could only base64 encode the WASM file to load it in Worklet context?

I dont known

> Are there any other APIs used by wasm-bindgen that is missing in Worklet context?

wasm-bindgen uses `TextEncoder` and `TextDecoder` [(see)](https://github.com/rustwasm/wasm-bindgen/issues/2367).

**Comment by Menci** [comment]
Firefox worker is supported now. Please test.

**Comment by Jerboas86** [comment]
worker is still not working on FF, for both svelte and sveltekit (tested with with vite-plugin-wasm last commit)

this error message appears in the console. 
```
SyntaxError: import declarations may only appear at top level of a module
```

**Comment by Menci** [comment]
How did you build?

You need to set worker to iife, not es.

**Comment by Jerboas86** [comment]
> How did you build?

I retrieved the commit and built from source, maybe there is a better way?

package.json
```
{
  [...]
  "devDependencies": {
   [...]
    "vite-plugin-wasm": "git@github.com:Menci/vite-plugin-wasm.git#7ab4a5e"
  }
}
```

> You need to set worker to iife, not es.

i used the default as specify in the README

vite.config.js
```
import { defineConfig } from "vite";
import { svelte } from "@sveltejs/vite-plugin-svelte";
import wasm from "vite-plugin-wasm";
import topLevelAwait from "vite-plugin-top-level-await";

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [svelte(), wasm(), topLevelAwait()],
  test: {
    environment: "jsdom",
  },
  worker: {
    plugins: [wasm(), topLevelAwait()],
  },
});
```

I push it on the repo, you can check it out, see if there is any mistake. [(link)](https://github.com/Jerboas86/vpw-test-env)

**Comment by Menci** [comment]
No, you need to use my plugin vite-plugin-top-level-await to transform this.
